File: mocida/src/uikit/glass.c
#include <uikit/glass.h>
#include <string.h>
#include <stdlib.h>

// =====================================================================
// Material helpers (pure logic — same on every platform)
// =====================================================================

UIBackdropMaterial UIBackdrop_FromString(const char* effect) {
    if (!effect || !effect[0]) return UI_BACKDROP_AUTO;

    // Case-insensitive compare against a known set. Hyphens and underscores
    // are treated the same so "mica-alt" and "mica_alt" both work.
    char buf[32];
    size_t n = 0;
    for (const char* p = effect; *p && n < sizeof(buf) - 1; p++) {
        char c = *p;
        if (c >= 'A' && c <= 'Z') c = (char)(c - 'A' + 'a');
        if (c == '_') c = '-';
        buf[n++] = c;
    }
    buf[n] = '\0';

    if (!strcmp(buf, "auto"))             return UI_BACKDROP_AUTO;
    if (!strcmp(buf, "none") ||
        !strcmp(buf, "off"))              return UI_BACKDROP_NONE;
    if (!strcmp(buf, "mica"))             return UI_BACKDROP_MICA;
    if (!strcmp(buf, "mica-alt"))         return UI_BACKDROP_MICA_ALT;
    if (!strcmp(buf, "acrylic"))          return UI_BACKDROP_ACRYLIC;
    if (!strcmp(buf, "acrylic-legacy"))   return UI_BACKDROP_ACRYLIC_LEGACY;
    if (!strcmp(buf, "liquid") ||
        !strcmp(buf, "liquid-glass"))     return UI_BACKDROP_LIQUID_GLASS;
    if (!strcmp(buf, "vibrancy-sidebar")) return UI_BACKDROP_VIBRANCY_SIDEBAR;
    if (!strcmp(buf, "vibrancy-header"))  return UI_BACKDROP_VIBRANCY_HEADER;
    if (!strcmp(buf, "vibrancy-menu"))    return UI_BACKDROP_VIBRANCY_MENU;
    if (!strcmp(buf, "vibrancy-popover")) return UI_BACKDROP_VIBRANCY_POPOVER;
    if (!strcmp(buf, "vibrancy-hud"))     return UI_BACKDROP_VIBRANCY_HUD;
    if (!strcmp(buf, "kde-window"))       return UI_BACKDROP_KDE_BLUR_WINDOW;
    if (!strcmp(buf, "kde-region") ||
        !strcmp(buf, "kde-blur"))         return UI_BACKDROP_KDE_BLUR_REGION;

    return UI_BACKDROP_AUTO;
}

uint32_t UIBackdropMaterial_SupportedProps(UIBackdropMaterial m) {
    switch (m) {
        case UI_BACKDROP_MICA:
        case UI_BACKDROP_MICA_ALT:
            return UI_GLASS_PROP_TINT;

        case UI_BACKDROP_ACRYLIC:
        case UI_BACKDROP_ACRYLIC_LEGACY:
            return UI_GLASS_PROP_TINT | UI_GLASS_PROP_NOISE;

        case UI_BACKDROP_LIQUID_GLASS:
            return UI_GLASS_PROP_TINT | UI_GLASS_PROP_REFRACTION;

        case UI_BACKDROP_VIBRANCY_SIDEBAR:
        case UI_BACKDROP_VIBRANCY_HEADER:
        case UI_BACKDROP_VIBRANCY_MENU:
        case UI_BACKDROP_VIBRANCY_POPOVER:
            return UI_GLASS_PROP_STATE;

        case UI_BACKDROP_VIBRANCY_HUD:
            return UI_GLASS_PROP_STATE | UI_GLASS_PROP_NOISE;

        case UI_BACKDROP_KDE_BLUR_WINDOW:
        case UI_BACKDROP_KDE_BLUR_REGION:
            return UI_GLASS_PROP_TINT | UI_GLASS_PROP_BLUR;

        case UI_BACKDROP_NONE:
        case UI_BACKDROP_AUTO:
        default:
            // The in-app fallback understands everything it can paint.
            return UI_GLASS_PROP_TINT | UI_GLASS_PROP_BLUR | UI_GLASS_PROP_NOISE;
    }
}

int UIBackdropMaterial_IsWindowWide(UIBackdropMaterial m) {
    return m == UI_BACKDROP_MICA
        || m == UI_BACKDROP_MICA_ALT
        || m == UI_BACKDROP_KDE_BLUR_WINDOW;
}

// =====================================================================
// UIGlass widget (mirrors UIStack, plus the glass styling fields)
// =====================================================================

UIGlass* UIGlass_Create(UIBackdropMaterial material) {
    UIGlass* g = (UIGlass*)calloc(1, sizeof(UIGlass));
    if (!g) return NULL;
    g->__widget_type = UI_WIDGET_GLASS;
    g->material      = material;
    g->radius        = 0.0f;
    g->tint          = (UIColor){ 255, 255, 255, 1.0f };
    g->tintOpacity   = 0.5f;
    g->thickness     = UI_GLASS_REGULAR;
    g->refraction    = 0.0f;
    g->blurPx        = 16.0f;
    g->noise         = 0.0f;
    g->vibrancyState = UI_VIBRANCY_ACTIVE;
    g->orientation   = 0; // vertical
    g->spacing       = 8.0f;
    g->items         = UIChildren_Create(16);
    if (!g->items) { free(g); return NULL; }
    return g;
}

UIGlass* UIGlass_SetRadius(UIGlass* g, float radius) {
    if (g) g->radius = radius;
    return g;
}

UIGlass* UIGlass_SetTint(UIGlass* g, UIColor tint) {
    if (g) g->tint = tint;
    return g;
}

UIGlass* UIGlass_SetTintOpacity(UIGlass* g, float opacity) {
    if (g) g->tintOpacity = opacity;
    return g;
}

UIGlass* UIGlass_SetThickness(UIGlass* g, UIGlassThickness thickness) {
    if (g) g->thickness = thickness;
    return g;
}

UIGlass* UIGlass_SetBlur(UIGlass* g, float px) {
    if (g) g->blurPx = px;
    return g;
}

UIGlass* UIGlass_SetRefraction(UIGlass* g, float refraction) {
    if (g) g->refraction = refraction;
    return g;
}

UIGlass* UIGlass_SetNoise(UIGlass* g, float noise) {
    if (g) g->noise = noise;
    return g;
}

UIGlass* UIGlass_SetVibrancyState(UIGlass* g, UIVibrancyState state) {
    if (g) g->vibrancyState = state;
    return g;
}

UIGlass* UIGlass_SetOrientation(UIGlass* g, int horizontal) {
    if (g) g->orientation = horizontal ? 1 : 0;
    return g;
}

UIGlass* UIGlass_SetSpacing(UIGlass* g, float spacing) {
    if (g) g->spacing = spacing;
    return g;
}

UIGlass* UIGlass_SetAlign(UIGlass* g, int align) {
    if (g) g->align = align;
    return g;
}

UIGlass* UIGlass_SetJustify(UIGlass* g, int justify) {
    if (g) g->justify = justify;
    return g;
}

UIGlass* UIGlass_SetPadding(UIGlass* g, float l, float t, float r, float b) {
    if (!g) return g;
    g->paddingLeft   = l;
    g->paddingTop    = t;
    g->paddingRight  = r;
    g->paddingBottom = b;
    return g;
}

UIGlass* UIGlass_SetFreeLayout(UIGlass* g, int enabled) {
    if (g) g->freeLayout = enabled ? 1 : 0;
    return g;
}

int UIGlass_AddItem(UIGlass* g, UIWidget* item) {
    if (!g || !item) return 0;
    return UIChildren_Add(g->items, item);
}

void UIGlass_GetContentSize(UIGlass* g, float* outW, float* outH) {
    float w = 0.0f, h = 0.0f;
    if (g && g->items) {
        for (int i = 0; i < g->items->count; i++) {
            UIWidget* it = g->items->children[i];
            if (!it || !it->visible) continue;
            const float iw = it->width  ? *it->width  : 0.0f;
            const float ih = it->height ? *it->height : 0.0f;
            if (g->orientation == 1) {        // horizontal
                w += iw + (w > 0 ? g->spacing : 0.0f);
                if (ih > h) h = ih;
            } else {                          // vertical
                h += ih + (h > 0 ? g->spacing : 0.0f);
                if (iw > w) w = iw;
            }
        }
        w += g->paddingLeft + g->paddingRight;
        h += g->paddingTop  + g->paddingBottom;
    }
    if (outW) *outW = w;
    if (outH) *outH = h;
}

void UIGlass_Destroy(UIGlass* g) {
    if (!g) return;
    if (g->items) UIChildren_Destroy(g->items);
    free(g);
}

File: mocida-rs/mocida/src/glass.rs
//! `UIGlass` — region backdrop container (Acrylic / Liquid Glass / Vibrancy /
//! KDE blur), plus the [`BackdropMaterial`] family shared with the window-level
//! backdrop ([`Window::set_backdrop`](crate::window::Window::set_backdrop)).

use std::ffi::CString;

use mocida_sys as sys;

use crate::color::Color;
use crate::error::{Error, Result};
use crate::widget::Widget;

/// Backdrop material family. Mirrors the C `UIBackdropMaterial` enum 1:1.
///
/// `Mica` / `MicaAlt` / `KdeBlurWindow` are *window-wide* (use them with
/// [`Window::set_backdrop`](crate::window::Window::set_backdrop)); the rest are
/// *region* materials usable on a [`Glass`] widget.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum BackdropMaterial {
    /// No effect.
    None = 0,
    /// Platform default, resolved at runtime.
    Auto = 1,
    /// Win11 Mica (window-wide).
    Mica = 2,
    /// Win11 Mica Alt (window-wide).
    MicaAlt = 3,
    /// KWin whole-window blur (window-wide).
    KdeBlurWindow = 4,
    /// Win11 Acrylic.
    Acrylic = 5,
    /// Win10 acrylic blur-behind.
    AcrylicLegacy = 6,
    /// macOS/iOS 26 Liquid Glass.
    LiquidGlass = 7,
    /// macOS sidebar vibrancy.
    VibrancySidebar = 8,
    /// macOS header vibrancy.
    VibrancyHeader = 9,
    /// macOS menu vibrancy.
    VibrancyMenu = 10,
    /// macOS popover vibrancy.
    VibrancyPopover = 11,
    /// macOS HUD vibrancy.
    VibrancyHud = 12,
    /// KWin per-region blur.
    KdeBlurRegion = 13,
}

impl BackdropMaterial {
    /// Parse an `effect:` string ("mica", "acrylic", "liquid",
    /// "vibrancy-sidebar", "kde-region", …) using the C resolver, so the
    /// mapping stays single-sourced. Unknown / empty → [`Auto`](Self::Auto).
    pub fn from_effect(effect: &str) -> Self {
        let c = match CString::new(effect) {
            Ok(c) => c,
            Err(_) => return BackdropMaterial::Auto,
        };
        let raw = unsafe { sys::UIBackdrop_FromString(c.as_ptr()) };
        Self::from_raw(raw.0)
    }

    /// 1:1 from the raw enum value (out-of-range → [`Auto`](Self::Auto)).
    pub fn from_raw(v: i32) -> Self {
        match v {
            0 => Self::None,
            1 => Self::Auto,
            2 => Self::Mica,
            3 => Self::MicaAlt,
            4 => Self::KdeBlurWindow,
            5 => Self::Acrylic,
            6 => Self::AcrylicLegacy,
            7 => Self::LiquidGlass,
            8 => Self::VibrancySidebar,
            9 => Self::VibrancyHeader,
            10 => Self::VibrancyMenu,
            11 => Self::VibrancyPopover,
            12 => Self::VibrancyHud,
            13 => Self::KdeBlurRegion,
            _ => Self::Auto,
        }
    }

    #[inline]
    pub(crate) fn raw(self) -> sys::UIBackdropMaterial {
        sys::UIBackdropMaterial(self as i32)
    }

    /// True for window-wide materials (Mica / Mica Alt / KDE blur-window).
    pub fn is_window_wide(self) -> bool {
        matches!(self, Self::Mica | Self::MicaAlt | Self::KdeBlurWindow)
    }
}

/// Material thickness preset (mirrors `UIGlassThickness`).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum GlassThickness {
    /// Thin material.
    Thin = 0,
    /// Default material.
    Regular = 1,
    /// Thick material.
    Thick = 2,
}

/// AppKit vibrancy state (mirrors `UIVibrancyState`).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum VibrancyState {
    /// Active appearance.
    Active = 0,
    /// Inactive appearance.
    Inactive = 1,
    /// Pressed appearance.
    Pressed = 2,
}

/// Region-backdrop container. Lays children like a vertical [`Stack`](crate::Stack)
/// while drawing a glass background (native OS effect where available, in-app
/// tint otherwise).
pub struct Glass {
    ptr: *mut sys::UIGlass,
    moved: bool,
}

impl Glass {
    /// Creates a glass container with the given material.
    pub fn new(material: BackdropMaterial) -> Result<Self> {
        let ptr = unsafe { sys::UIGlass_Create(material.raw()) };
        if ptr.is_null() {
            return Err(Error::Null("UIGlass_Create"));
        }
        Ok(Self { ptr, moved: false })
    }

    /// Corner radius of the glass fill.
    pub fn radius(self, radius: f32) -> Self {
        unsafe { sys::UIGlass_SetRadius(self.ptr, radius) };
        self
    }

    /// Overlay tint color.
    pub fn tint(self, color: Color) -> Self {
        unsafe { sys::UIGlass_SetTint(self.ptr, color.into_raw()) };
        self
    }

    /// Tint strength, `0.0..=1.0`.
    pub fn tint_opacity(self, opacity: f32) -> Self {
        unsafe { sys::UIGlass_SetTintOpacity(self.ptr, opacity) };
        self
    }

    /// Material thickness preset.
    pub fn thickness(self, thickness: GlassThickness) -> Self {
        unsafe { sys::UIGlass_SetThickness(self.ptr, sys::UIGlassThickness(thickness as i32)) };
        self
    }

    /// KDE blur radius in pixels (ignored by materials that lack a blur knob).
    pub fn blur(self, px: f32) -> Self {
        unsafe { sys::UIGlass_SetBlur(self.ptr, px) };
        self
    }

    /// Liquid Glass refraction, `0.0..=1.0` (ignored by other materials).
    pub fn refraction(self, refraction: f32) -> Self {
        unsafe { sys::UIGlass_SetRefraction(self.ptr, refraction) };
        self
    }

    /// Acrylic / HUD grain, `0.0..=1.0` (ignored by other materials).
    pub fn noise(self, noise: f32) -> Self {
        unsafe { sys::UIGlass_SetNoise(self.ptr, noise) };
        self
    }

    /// AppKit vibrancy state (ignored by non-vibrancy materials).
    pub fn vibrancy_state(self, state: VibrancyState) -> Self {
        unsafe { sys::UIGlass_SetVibrancyState(self.ptr, sys::UIVibrancyState(state as i32)) };
        self
    }

    /// Lay children left-to-right (`true`) instead of top-to-bottom.
    pub fn horizontal(self, horizontal: bool) -> Self {
        unsafe { sys::UIGlass_SetOrientation(self.ptr, horizontal as i32) };
        self
    }

    /// Gap between consecutive items.
    pub fn spacing(self, spacing: f32) -> Self {
        unsafe { sys::UIGlass_SetSpacing(self.ptr, spacing) };
        self
    }

    /// Cross-axis alignment (0 start, 1 center, 2 end).
    pub fn align(self, align: i32) -> Self {
        unsafe { sys::UIGlass_SetAlign(self.ptr, align) };
        self
    }

    /// Main-axis distribution (0 start, 1 center, 2 end, 3 between).
    pub fn justify(self, justify: i32) -> Self {
        unsafe { sys::UIGlass_SetJustify(self.ptr, justify) };
        self
    }

    /// Inner padding (left, top, right, bottom).
    pub fn padding(self, left: f32, top: f32, right: f32, bottom: f32) -> Self {
        unsafe { sys::UIGlass_SetPadding(self.ptr, left, top, right, bottom) };
        self
    }

    /// Absolute-position children instead of flowing them.
    pub fn free_layout(self, enabled: bool) -> Self {
        unsafe { sys::UIGlass_SetFreeLayout(self.ptr, enabled as i32) };
        self
    }

    /// Appends an item. The glass takes ownership of the widget.
    pub fn add(&mut self, item: Widget) -> Result<()> {
        let raw = item.into_raw();
        let rc = unsafe { sys::UIGlass_AddItem(self.ptr, raw) };
        if rc != 1 {
            return Err(Error::Null("UIGlass_AddItem"));
        }
        Ok(())
    }

    /// Borrow the raw `UIGlass*`.
    #[inline]
    pub fn as_ptr(&self) -> *mut sys::UIGlass {
        self.ptr
    }

    /// Lift into a [`Widget`] (auto-sized).
    pub fn into_widget(mut self) -> Result<Widget> {
        self.moved = true;
        unsafe { Widget::new(self.ptr as *mut _) }
    }

    /// Lift into a [`Widget`] with an explicit size.
    pub fn into_widget_sized(mut self, width: f32, height: f32) -> Result<Widget> {
        self.moved = true;
        unsafe { Widget::with_size(self.ptr as *mut _, width, height) }
    }
}

impl Drop for Glass {
    fn drop(&mut self) {
        if !self.moved && !self.ptr.is_null() {
            unsafe { sys::UIGlass_Destroy(self.ptr) };
        }
    }
}

/// Global glass switch (live, no rebuild). When `false`, every `Glass` widget
/// renders as a SOLID panel (full-opacity tint, no blur) — so a `Glass{}` can
/// double as a plain panel and the app toggles "translucent" at runtime.
pub fn set_glass_enabled(on: bool) {
    unsafe { sys::UIGlass_SetGlobalEnabled(if on { 1 } else { 0 }) };
}

/// Global blur-strength multiplier applied to every `Glass` widget's `blur`
/// (the intensity / "power" knob). 1.0 = as authored.
pub fn set_glass_blur_scale(scale: f32) {
    unsafe { sys::UIGlass_SetGlobalBlurScale(scale) };
}
